storage/uri.go

```diff
@@ -59,53 +61,131 @@ func (u *uri) String() string {
 	return u.raw
 }
 
+// parentGeneric is a generic function that returns the last element of a
+// path after splitting it on "/". It should be suitable for most URIs.
+func parentGeneric(location string) (string, error) {
+
+	// trim leading forward slashes
+	trimmed := 0
+	for location[0] == '/' {
+		location = location[1:]
+		trimmed++
+
+		// if all we have left is an empty string, than this URI
+		// pointed to a UNIX-style root
+		if len(location) == 0 {
+			return "", URIRootError
+		}
+	}
+
+	components := strings.Split(location, "/")
+
+	if len(components) == 1 {
+		// trimmed <= 2 makes sure we handle UNIX-style paths on
+		// Windows correctly
+		return "", URIRootError
+	}
+
+	parent := ""
+	if trimmed > 2 && len(components) > 1 {
+		// Because we trimmed all the leading '/' characters, for UNIX
+		// style paths we want to insert one back in. Presumably we
+		// trimmed two instances of / for the scheme.
+		parent = parent + "/"
+	}
+	parent = parent + strings.Join(components[0:len(components)-1], "/") + "/"
+
+	return parent, nil
+}
+
 // Parent gets the parent of a URI by splitting it along '/' separators and
 // removing the last item.
 func Parent(u fyne.URI) (fyne.URI, error) {
 	s := u.String()
 
 	// trim trailing slash
+	if runtime.GOOS == "windows" {
+		if s[len(s)-1] == '\\' {
+			s = s[0 : len(s)-1]
+		}
+	}
 	if s[len(s)-1] == '/' {
 		s = s[0 : len(s)-1]
 	}
 
-	// trim the scheme (and +1 for the :)
-	s = s[len(u.Scheme())+1:]
+	// trim the scheme
+	s = s[len(u.Scheme())+3:]
```

Comment:
Why +3? the comment about the +1 has gone so it's not clear?
If it's for :// then beware that the "//" is optional in URI spec

Reply:
> beware that the "//" is optional in URI spec

We are using ```len(Scheme())+3``` everywhere. There should be some functions like ```Path()``` which returns only the path of an URI to be complient with that.

Reply:
But this code is improving the very way that URIs are being parsed and created - so we should be trying to improve rather than replicating the same issues in the previous implementations :).

The reason that Path() was not added immediately iis that if people thought this was just about files then they might do os.File operations on the result of URI.Path(), which is often not going to work.
In the future we can add it, once we're cleanly using abstractions in our own code. But this is a breaking change as the URI interface is public so it would not happen before 2.0.

Reply:
What about ```path()``` for internal use?

Reply:
I have pitched the idea of having a standardized method for retrieving the parts of the URI that aren't the scheme in the past. I still think it would be a good ideas as @PucklaMotzer09 mentioned, even if it was only for internal use. We could also put it in `storage` like `Parent` for now, if we wanted it to be public.